presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ApplyConnectorOptimization.java

```diff
@@ -0,0 +1,260 @@
+
+            // keep track a set of changed nodes; the keys are original nodes and the values are the new nodes
+            Map<PlanNode, PlanNode> updates = new HashMap<>();
+
+            // process connector optimizers
+            for (PlanNode node : contextMap.keySet()) {
+                // get max closure
+                ConnectorPlanNodeContext context = contextMap.get(node);
+                if (!context.isClosure(connectorId) ||
+                        !context.getParent().isPresent() ||
+                        contextMap.get(context.getParent().get()).isClosure(connectorId)) {
+                    continue;
+                }
+
+                PlanNode newNode = node;
+
+                // TODO: the returned node is still a max closure only if there is no new connector added
+                for (ConnectorPlanOptimizer optimizer : optimizers) {
+                    newNode = optimizer.optimize(newNode, connectorSession, symbolAllocator, idAllocator);
+                }
+
+                if (node != newNode) {
+                    // the optimizer does allocate a new PlanNode
+                    updates.put(node, newNode);
+                }
+            }
+            // up to this point, we have a set of updated nodes; need to recursively update their parents
+
+            // alter the plan with a bottom-up approach
+            // use "original nodes" to keep track of the plan structure and "updates" to keep track of the new nodes
+            Queue<PlanNode> originalNodes = new LinkedList<>(updates.keySet());
+            while (!originalNodes.isEmpty()) {
+                PlanNode originalNode = originalNodes.poll();
+
+                if (!contextMap.get(originalNode).getParent().isPresent()) {
+                    // originalNode must be the root; update the plan
+                    plan = updates.get(originalNode);
+                    continue;
+                }
+
+                PlanNode originalParent = contextMap.get(originalNode).getParent().get();
+
+                // need to create a new parent given the child has changed; the new parent needs to point to the new child.
+                // if a node has been updated, it will occur in `updates`; otherwise, just use the original node
+                ImmutableList.Builder<PlanNode> newChildren = ImmutableList.builder();
+                originalParent.getSources().forEach(child -> newChildren.add(updates.getOrDefault(child, child)));
+                PlanNode newParent = originalParent.replaceChildren(newChildren.build());
+
+                // mark the new parent as updated
+                updates.put(originalParent, newParent);
+
+                // enqueue the parent node in order to recursively update its ancestors
+                originalNodes.add(originalParent);
+            }
+        }
+
+        return plan;
+    }
+
+    private static Set<ConnectorId> getAllConnectorIds(PlanNode node)
```

Comment:
another pattern in Presto is to allow a modifiable parameter, e.g. 

```
void retrieveAllConnectorIds(PlanNode node, ImmutableList.Builder connectorIds)
```

For example, see https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java#L390-L391


I don't have strong preference which way to go here :)